### ISSUE TITLE
Heatmap support on ElasticSearch

### ISSUE THREAD

**Issue by panchicore** [cga-harvard/HHypermap#204]
docker hub terranodo/elasticsearch:heatmap

https://github.com/o19s/elasticsearch/tree/feature/heatmap

**Comment by panchicore** [comment]
I think I am having errors related with the geoheatmap work.

```
...elasticsearch/core/src/main/java/org/elasticsearch/search/aggregations/bucket/geoheatmap/GeoHeatmapAggregationBuilder.java:97: warning: no description for @throws
     * @throws IOException
       ^
error: warnings found and -Werror specified
Note: Some input files use or override a deprecated API.
Note: Recompile with -Xlint:deprecation for details.
Note: Some input files use unchecked or unsafe operations.
Note: Recompile with -Xlint:unchecked for details.
1 error
4 warnings
:core:compileJava FAILED
```

looking into it.

**Comment by panchicore** [comment]
they made some fixes [here](https://github.com/o19s/elasticsearch/commit/2ce5376b4fd0071dd8b6868988bbb29e6b144dd9) but not applied to the `GeoHeatmapAggregationBuilder.java` class. commenting code btw `GeoHeatmapAggregationBuilder.java:94-98`.

**Comment by panchicore** [comment]
@ingenieroariel just wondering why:

master has a folder: geoheatmap:

https://github.com/o19s/elasticsearch/tree/master/core/src/main/java/org/elasticsearch/search/aggregations/bucket

feature/heatmap not:

https://github.com/o19s/elasticsearch/tree/feature/heatmap/core/src/main/java/org/elasticsearch/search/aggregations/bucket

**Comment by ingenieroariel** [comment]
I'll point the main author of the feature branch at this thread. Thanks
Luis.

**Comment by sstults** [comment]
@panchicore, we had been collecting document "hits" on the heatmap (so that would be a bucket aggregation), but since the Lucene heatmap doesn't keep track of this at all there's no reason for us to do that, and there's a margin for error.

Instead we decided to use the metrics aggregation type, and so we moved the classes from the bucket package to the metrics package:

https://github.com/o19s/elasticsearch/tree/feature/heatmap/core/src/main/java/org/elasticsearch/search/aggregations/metrics/geoheatmap

**Comment by panchicore** [comment]
@sstults good to know. 

@sstults do you have query samples about how to filter with that aggregation type?

our docs has this mapping:

```
mapping = {
            "mappings": {
                "layer": {
                    "properties": {
                        "layer_geoshape": {
                           "type": "geo_shape",
                           "tree": "quadtree",
                           "precision": REGISTRY_MAPPING_PRECISION
                        }
                    }
                }
            }
        }
```

so we will be creating heatmaps from the `layer_geoshape` geo shape.

I was looking into the documentation and found [geohash aggregations](https://www.elastic.co/guide/en/elasticsearch/reference/5.x/search-aggregations-bucket-geohashgrid-aggregation.html) but in my tests it only works with `geo_point` type, not with the `geo_shape` one.

but I realize that maybe I will find the implementation by looking in the geoheatmap tests: https://github.com/o19s/elasticsearch/tree/feature/heatmap/core/src/test/java/org/elasticsearch/search/aggregations/metrics/geoheatmap

thanks in advance.

**Comment by sstults** [comment]
Correct - there aren't any aggregations so far that support geo_shape types.

Right now we've got two layers to restrict the bounds on the heatmap. The first is through normal query/filter parameters. So if you execute a regular geo_shape query or filter and look at the resulting documents, those are what would be passed to the heatmap agg.

The heatmap itself is another way to restrict what gets into the heatmap. There's a "geom" parameter that takes a geo_shape query as another filter. So for example you could do a top-level query for things in a country, and then use the geom field in the agg to restrict to a particular city.

Something like this:

```javascript
{
    "query": {
        "geo_shape": {
            "location": {
                "shape": {
                    "type": "envelope",
                    "coordinates" : [[-50.0, 43.0], [50.0, 44.0]]
                },
                "relation": "within"
            }
        }
    },
    "aggs" : {
        "viewport" : {
            "heatmap" : {
                "field" : "location", 
                "grid_level" : 4,
                "max_cells" : 100,
                "geom" : {
                    "geo_shape": {
                        "location": {
                            "shape": {
                                "type": "envelope",
                                "coordinates" : [[-30.0, 30.0], [40.0, -30.0]]
                            },
                            "relation": "within"
                        }
                    }
                }
            }
        }
    }
}
```

The actual coordinates above are silly, but hopefully you get the idea.

**Comment by panchicore** [comment]
with 350 layers where the majority covers entire earth.

the following is the query I am running against [es5.0 feature/heatmap](https://github.com/o19s/elasticsearch/tree/feature/heatmap): 

```json
{
    "sort":{
        "_score":{
            "order":"desc"
        }
    },
    "query":{
        "bool":{
            "filter":{
                "geo_shape":{
                    "layer_geoshape":{
                        "shape":{
                            "type":"envelope",
                            "coordinates":[
                                [
                                    "-180.0",
                                    "90.0"
                                ],
                                [
                                    "180.0",
                                    "-90.0"
                                ]
                            ]
                        },
                        "relation":"intersects"
                    }
                }
            }
        }
    },
    "from":0,
    "aggs":{
        "viewport":{
            "heatmap":{
                "field":"layer_geoshape",
                "geom":{
                    "geo_shape":{
                        "layer_geoshape":{
                            "shape":{
                                "type":"envelope",
                                "coordinates":[
                                    [
                                        "-180.0",
                                        "90.0"
                                    ],
                                    [
                                        "180.0",
                                        "-90.0"
                                    ]
                                ]
                            },
                            "relation":"intersects"
                        }
                    }
                },
                "grid_level":3,
                "max_cells":18432
            }
        }
    }
}

```

getting as result:

```json

{
    "a.hm":{
        "minX":-180.0,
        "minY":-90.0,
        "maxX":180.0,
        "rows":8,
        "projection":"EPSG:4326",
        "counts_ints2D":[
            [
                [
                    350,
                    311,
                    342,
                    350,
                    350,
                    350,
                    350,
                    350
                ]
            ],
            [
                [
                    350,
                    311,
                    342,
                    350,
                    350,
                    350,
                    350,
                    350
                ]
            ],
            [
                [
                    350,
                    311,
                    342,
                    350,
                    350,
                    350,
                    350,
                    350
                ]
            ],
            [
                [
                    322,
                    311,
                    314,
                    322,
                    322,
                    322,
                    322,
                    322
                ]
            ],
            [
                [
                    320,
                    309,
                    312,
                    320,
                    320,
                    320,
                    320,
                    320
                ]
            ],
            [
                [
                    316,
                    309,
                    310,
                    316,
                    316,
                    316,
                    316,
                    316
                ]
            ],
            [
                [
                    309,
                    309,
                    309,
                    309,
                    309,
                    309,
                    309,
                    309
                ]
            ],
            [
                [
                    309,
                    309,
                    309,
                    309,
                    309,
                    309,
                    309,
                    309
                ]
            ]
        ],
        "gridLevel":3,
        "maxY":180.0,
        "columns":8
    },
...
}
```

(thats a transformed response from [the original response](http://hastebin.com/lesegeboke.rb))

the findings in that response versus the solr response with the same 350 layers are:

- ES returns rows:8 colums:8 square. Solr returns rows:4 colums:8 rect.
- ES was set to gridLevel=3. Solr was set to 1. ES with 1 returned a rows:2 columns:2 matrix, so I incremented the gridLevel to get nearest result to Solr.
- columns: Solr values are contained in ES values. 

es[0]==solr[0]
es[1]==solr[1]
es[4]==solr[2]
es[5]==solr[3]

here Solr results:

```json

{
    "a.hm":{
        "minX":-180.0,
        "minY":-90.0,
        "maxX":180.0,
        "rows":4,
        "projection":"EPSG:4326",
        "counts_ints2D":[
            [
                350,
                311,
                342,
                350,
                350,
                350,
                350,
                350
            ],
            [
                350,
                311,
                342,
                350,
                350,
                350,
                350,
                350
            ],
            [
                320,
                309,
                312,
                320,
                320,
                320,
                320,
                320
            ],
            [
                309,
                309,
                309,
                309,
                309,
                309,
                309,
                309
            ]
        ],
        "gridLevel":1,
        "maxY":90.0,
        "columns":8
    }
}

```

@ingenieroariel @sstults  thoughts?

@ingenieroariel reference for us:

solr: `http://localhost:8000/registry/hypermap/api/?a.hm.gridlevel=1&a.hm.limit=1&a.text.limit=0&a.time.gap=P1D&a.time.limit=0&a.user.limit=0&d.docs.limit=0&d.docs.page=1&d.docs.sort=score&format=json&original_response=0&q.geo=%5B-90%2C-180+TO+90%2C180%5D&q.time=%5B1900-01-01+TO+2016-12-31T00%3A00%3A00%5D`

es: `http://localhost:8000/registry/hypermap/api/?a.hm.gridlevel=3&a.hm.limit=1&a.text.limit=0&a.time.gap=P1D&a.time.limit=0&a.user.limit=0&d.docs.limit=0&d.docs.page=1&d.docs.sort=score&format=json&original_response=0&q.geo=%5B-90%2C-180+TO+90%2C180%5D&q.time=%5B1900-01-01+TO+2016-12-31T00%3A00%3A00%5D`

**Comment by sstults** [comment]
It could be a bug in how we're converting the counts to JSON -- that's the very last functionality I added. The internal array should be the same since I'm doing exactly the same thing as Solr. One difference could be in the underlying field definition. Can you tell me how layer_geoshape is defined in both Solr and ES? 

By the way, score doesn't have meaning in strictly geo_shape queries so sorting might use memory and time unnecessarily. But if you're combining the geo_shape query with some sort of text query then score will matter.

**Comment by panchicore** [comment]
@sstults,

in Solr we declare the field like this:

```json
{"name": "bbox", "type": "location_rpt"},
```

that is what shows the eschema explorer:

<img src="http://panchicore.d.pr/FcO1+" width="500px">

`Field-Type:org.apache.solr.schema.SpatialRecursivePrefixTreeFieldType`


with ES we have this mapping at the moment of the indice creation:

```
"mappings": {
                "layer": {
                    "properties": {
                        "layer_geoshape": {
                           "type": "geo_shape",
                           "tree": "quadtree",
                           "precision": REGISTRY_MAPPING_PRECISION
                        }
                    }
                }
            }
```
`http://localhost:9200/hypermap/_mapping/layer` shows

```json
"layer_geoshape":{
                        "type":"geo_shape",
                        "tree":"quadtree",
                        "precision":"500.0m"
                    }

```

[that is the complete response json.](http://hastebin.com/elubesiqoc.json)

`solr SpatialRecursivePrefixTreeFieldType == es geo_shape quadtree` ?

what do you think? 

---
btw, thanks for the annotation about the score sorting.

**Comment by sstults** [comment]
@panchicore, check out minX, minY, maxX, and maxY between the two responses. I think you need to switch the coordinates. I'm looking for a good reference on GeoJSON now, but the only thing I've found is this, and it doesn't describe the coordinate order for Envelope:

http://wiki.geojson.org/GeoJSON_draft_version_5#Including_Bounding_Box.2FEnvelope

**Comment by panchicore** [comment]
@sstults,

ENVELOPE Catalogue Service Specification, I know because this thread [here](https://github.com/locationtech/spatial4j/issues/36#issuecomment-28917634).

What happens if I switch the coords?

```json

{

    "query":{
        "bool":{
            "filter":{
                "geo_shape":{
                    "layer_geoshape":{
                        "shape":{
                            "type":"envelope",
                            "coordinates":[
                                [
                                    "-180.0",
                                    "-90.0"
                                ],
                                [
                                    "180.0",
                                    "90.0"
                                ]
                            ]
                        },
                        "relation":"intersects"
                    }
                }
            }
        }
    },
    "from":0,
    "aggs":{
        "viewport":{
            "heatmap":{
                "field":"layer_geoshape",
                "geom":{
                    "geo_shape":{
                        "layer_geoshape":{
                            "shape":{
                                "type":"envelope",
                                "coordinates":[
                                    [
                                        "-180.0",
                                        "-90.0"
                                    ],
                                    [
                                        "180.0",
                                        "90.0"
                                    ]
                                ]
                            },
                            "relation":"intersects"
                        }
                    }
                },
                "grid_level":3,
                "max_cells":18432
            }
        }
    }
}
```

the response is exactly the same.

in case you want to dump the 350 layers, you can download them [here](http://panchicore.d.pr/19lBg). I got it with `http://localhost:9200/hypermap/_search?size=400`

what do you think?

**Comment by panchicore** [comment]
@sstults I will prepare a test script for you to know about the responses expected from some queries taking as base reference the solr responses. WIP.

**Comment by sstults** [comment]
@panchicore, I think I found the error. Have a look at the change I just posted in the patch to views.py.

**Comment by panchicore** [comment]
ok. I will check and let you know today.

https://github.com/sstults/HHypermap/commit/429a82985f50baa4511cb32d79d5d6fed41d6457

**Comment by panchicore** [comment]
@sstults thanks for the feedback. You found an error when the ES output is extracted to match with the response model, that error does not affect: 

1. the input: the data sent to ES is the same that is being sent to Solr. 
2. the throughput: the process is based in the information here:  https://github.com/sstults/HHypermap/blob/429a82985f50baa4511cb32d79d5d6fed41d6457/hypermap/search_api/views.py#L261 where geo shape envelope coords are set OK.

Anyways, I re runned the tests and your finding is not conclusive and does not affect the results of my prev comments.

I will send you a python script with some tests in order to you know what we spec for ES.

thanks in advance,

Luis.

**Comment by sstults** [comment]
Correct. I was concerned that maybe the bounding box coordinates weren't being sent in the right order, but it looks like the error was just in the output.

With respect to the extra column, I ran into something similar in my unit test and that was actually an _indexing_ problem. I was reproducing a Lucene test that indexed an envelope like this:

```
ENVELOPE(100, 120, 80, 40)
```

My mistake was in taking the values in pairs:
```
"shape": {
                            "type": "envelope",
                            "coordinates" : [[100, 120], [80, 40]]
                }
```

When instead Solr puts the longitudes first, followed by the latitudes. So in ES my envelope should be:

```
"shape": {
                            "type": "envelope",
                            "coordinates" : [[100, 80], [120, 40]]
                }
```

It looks like the agg is built correctly in views.py, but I'm wondering if the indexing code mixes up the value order. Can you point me to where that would be?

Anyway, when I corrected the value order in the indexed docs in my own test I ended up with exactly the same output as the core Solr unit test, so that's encouraging.

**Comment by sstults** [comment]
@panchicore, I've submitted the PR to the main ES project based off of a copy of the code we've been working on. Everything's the same except there's no CircleCI and I added another unit test. Here's the fork I submitted:
https://github.com/sstults/elasticsearch

And here are the [PR](https://github.com/elastic/elasticsearch/pull/20666) and [Ticket](https://github.com/elastic/elasticsearch/issues/20665)

I'm not really sure whether we should keep working off of the o19s fork or the sstults fork, or if we can use the actual PR. Do you have a suggestion?

**Comment by panchicore** [comment]
@sstults I dont know what suggestion give. I am just compiling those sources (https://github.com/o19s/elasticsearch/tree/feature/heatmap) and running my tests. so, you mean I must clone your repo to make the tests?

**Comment by panchicore** [comment]
@sstults we need pick the grid level indirectly; we dont always specify a heatmap grid level, we calculate the distErr from the viewport rectangle and provide to solr a `facet.heatmap.distErr` parameter. see here: https://github.com/cga-harvard/HHypermap/blob/master/hypermap/search_api/utils.py#L279

Any posibility to make it work in your branch?

**Comment by ingenieroariel** [comment]
@panchicore let's move our build to use the branch on this repo: https://github.com/sstults/elasticsearch

For now let's use a fixed grid level if distErr is not supported.

**Comment by dsmiley** [comment]
That particular formula was, I believe, lifted from my HHypermap BOP web-service (Kotlin code)?  I think it's only valid for a quad-tree, by the way.  Consider it a best-effort estimate.

**Comment by panchicore** [comment]
thats an update from the heatmap implementation in both backends.

<img src="http://panchicore.d.pr/15J1n+">

I set gridlevel 7 in elasticsearch because I wanted to match the heatmap intensity with the solr one.

@dsmiley @sstults the gridlevel integer, its a standart value? I mean, both backends should give me the same gridlevel from exactly the same query?

If I manualy set gridlevel 7 to elasticsearch, it will result a less intense heatmap, what I think looks better.

<img src="http://panchicore.d.pr/u1cr+">

see the difference btw `counts_ints2D` and the `columns` count. why are so different but the map is drawing the same heatmap?

**Comment by sstults** [comment]
@ingenieroariel and @panchicore: I added distErr and distErrPct parameters to the heatmap as they're implemented in Solr, and in testing I've found something odd which may have the same root cause as the column disparity above.

The unit test I made follows the same logic as the Solr unit test and uses the same parameter values. However, when I use `geom` + `dist_err_pct` parameters, and when I use `geom` + `dist_err_pct` parameters, I'm getting a computed `grid_level` that's one less than expected. When I use the default `geom` though (the world rect) and `dist_err_pct` I get the expected value.

There are a couple of units conversions that could be throwing things off. The default units in Solr are degrees, whereas it looks like ES wants to use meters. I'm looking into whether or not this impacts the grid level calculation.

Would it be best for me to commit what I have now and follow up on the discrepancy at a later date?

**Comment by ingenieroariel** [comment]
I agree it'sbest to commit what's there and worry about the discrepancy later.

**Comment by ingenieroariel** [comment]
We can switch Solr to use meters instead I believe without much trouble.
